### PR TITLE
Source backoff strategy on failed events

### LINKED ISSUE #537: There is no failover or protection when events fail to be sent from the RabbitMQ Source
**Describe the bug**
When there is an error sending the events from the Source to a Sink they just keeps getting reqeued and RabbitMQ goes crazy:

screenshot_2021-12-02_at_15 58 22

When the peak in message rate in the graph starts to go down is when we configured the Knative DNS in the Cluster.

 **Expected behavior**
When message are not been sent, that the Source stops and wait for the Sink to be available. And maybe that the source (if the sink is a Ref) checks that a resolvable Ref.

 **To Reproduce**
- Spun up a Cluster environment with knative, but do not configure the DNS
- Create a RabbitMQ cluster
- Create a Source with a sink (uri or ref)
- Create the sink (optional)
- Get to the RabbitMQ Management UI and see how the image above gets replicated

 **Knative release version**
1.0.0

### PR DESCRIPTION
**Pull request by gabo1208**
# Changes

- :gift: Now using CE http retries when message fails
- :gift: There is a backoff strategy when retrying
- 🧹 Not reqeueing when prefetchCount is 1 (FIFO)

/kind performance

Fixes #537 #504

**Release Note**

```release-note
Now the RabbitMQ Source, when failing to send an event, does not get stuck in a cycle requeueing messages as they come. It follows a retries based backoff strategy (and does not reqeue them if the prefetchCount is 1)
```

### CODE REVIEW

File: pkg/apis/sources/v1alpha1/rabbitmq_types.go [hunk 1]
```diff
@@ -124,6 +125,22 @@ type RabbitmqSourceSpec struct {
 	// RabbitMQ cluster
 	// +optional
 	Predeclared bool `json:"predeclared,omitempty"`
+	// Retry is the minimum number of retries the sender should attempt when
+	// sending an event before moving it to the dead letter sink.
+	// +optional
+	Retry *int32 `json:"retry,omitempty"`
```
Comment (xtreme-sameer-vohra) [inline pkg/apis/sources/v1alpha1/rabbitmq_types.go:131]: Would this and `BackoffPolicy`,  `BackoffDelay` be available for users to configure or they'd have fixed values in code but wouldn't be configurable by users ?

Comment (gabo1208) [reply pkg/apis/sources/v1alpha1/rabbitmq_types.go:131]: After what we talked about today, I think this makes sense (they are configurable by users but have a default). Also the Eventing Source Spec "explicitly permit" additional data in the Source's Spec and Status (https://github.com/knative/eventing/blob/main/docs/spec/sources.md#source-custom-objects)

File: source/source-example.yaml [hunk 2]
```diff
@@ -6,6 +6,9 @@ metadata:
 spec:
   brokers: RABBITMQ_BROKERS
   topic: RABBITMQ_TOPIC
+  retry: 5
+  backoffPolicy: "linear"
+  backoffDelay: "PT30S"
```
Comment (gab-satchi) [inline source/source-example.yaml:11]: Does the period format "PT30S" work with time.Duration?

**Review by gab-satchi — COMMENTED** [review]
mostly looks good to me. just a few nits.

File: source/source-example.yaml (hunk 2 above)
Comment (gabo1208) [reply source/source-example.yaml:11]: You are right let me change it

**Review by gab-satchi — COMMENTED** [review]
/lgtm

**Comment by gab-satchi** [comment]
@gabo1208 maybe a squash to clean up the commits a bit :)

**Pull request merged**